tensorboard/webapp/plugins/text/text_v2_plugin.py

```diff
@@ -0,0 +1,215 @@
+from tensorboard.util import tensor_util
+from tensorboard.backend import http_util
+from tensorboard.plugins import base_plugin
+from tensorboard.data import provider
+from tensorboard.plugins.text import metadata
+
+# HTTP routes
+TAGS_ROUTE = "/tags"
+TEXT_ROUTE = "/text"
+
+WARNING_TEMPLATE = textwrap.dedent(
+    """\
+  **Warning:** This text summary contained data of dimensionality %d, but only \
+  2d tables are supported. Showing a 2d slice of the data instead."""
+)
+
+_DEFAULT_DOWNSAMPLING = 100  # text tensors per time series
+
+
+def reduce_to_2d(arr):
+    """Given a np.npdarray with nDims > 2, reduce it to 2d.
+
+    It does this by selecting the zeroth coordinate for every dimension greater
+    than two.
+
+    Args:
+      arr: a numpy ndarray of dimension at least 2.
+
+    Returns:
+      A two-dimensional subarray from the input array.
+
+    Raises:
+      ValueError: If the argument is not a numpy ndarray, or the dimensionality
+        is too low.
+    """
+    if not isinstance(arr, np.ndarray):
+        raise ValueError("reduce_to_2d requires a numpy.ndarray")
+
+    ndims = len(arr.shape)
+    if ndims < 2:
+        raise ValueError("reduce_to_2d requires an array of dimensionality >=2")
+    # slice(None) is equivalent to `:`, so we take arr[0,0,...0,:,:]
+    slices = ([0] * (ndims - 2)) + [slice(None), slice(None)]
+    return arr[slices]
+
+
+def reduce_and_jsonify(text_arr):
+    """Take a numpy.ndarray containing strings, and convert it into a
+    json-compatible list, also squashing it to two dimensions if necessary.
+
+    If the ndarray contains a single scalar string, then that ndarray is
+    converted to a list. If it contains an array of strings,
+    that array is converted to a list.  If the array contains dimensionality
+    greater than 2, all but two of the dimensions are removed, and a warning
+    message is generated.  The list, shape of the array, and any warning
+    message is returned.
+
+    Args:
+      text_arr: A numpy.ndarray containing strings.
+
+    Returns:
+        The JSON-compatible list
+        The shape of the array
+        A warning message (empty string if no warning)
+    """
+    warning = ""
+    if not text_arr.shape:
+        # It is a scalar. No need to put it in a table, just apply markdown
+        return text_arr.tolist(), text_arr.shape, warning
+    if len(text_arr.shape) > 2:
+        warning = plugin_util.markdown_to_safe_html(
+            WARNING_TEMPLATE % len(text_arr.shape)
+        )
+        text_arr = reduce_to_2d(text_arr)
+    return text_arr.tolist(), text_arr.shape, warning
+
+
+def create_event(wall_time, step, string_ndarray):
+    """Convert a text event into a JSON-compatible response with rank <= 2"""
+    formatted_string_array, shape, warning = reduce_and_jsonify(string_ndarray)
+    return {
+        "wall_time": wall_time,
+        "step": step,
+        "text": formatted_string_array,
+        "shape": shape,
+        "warning": warning,
```

Comment:
warning should be formed and presented in the FE (for internationalization and etc...). Please omit.

You should be able to detect the error by inspecting the shape and the `string_array`. If you think FE should not know about this, then, please change this to a boolean property that reads more like `has_partial_data`.

Reply:
right now there would be no way to detect on the frontend whether the `forrmatted_string_array` was squashed or not (because shape reflects its post-squashing shape, not its pre-squash shape).  I will change warning to a boolean.

Reply:
Ah, can we also have the raw shape? I think we can have better warning message if we had the original shape, too.

Reply:
Also, the field "warning" is quite ambiguous. I know the http_api is supposed to help with that but having self descriptive name would be better.

Reply:
sounds good, will include raw shape and will rename the `warning` field to something more descriptive

Reply:
done